File: experiments/benchmarks/ros/benchmark_python/benchmark_python/publisher.py
"""
Publisher node for ROS 2 Python benchmarking.

NOTE: This file was copied and slightly modified from the dora-rs benchmarks in 
https://github.com/dora-rs/dora-benchmark/blob/main/ros2/py_pubsub/setup.py
"""

import time

import numpy as np
import rclpy
from rclpy.node import Node
from std_msgs.msg import UInt64MultiArray

# NOTE: SIZES defines the number of uint64 elements in the payload.
# Since each uint64 is 8 bytes, the actual payload size in bytes is: Element Count * 8.
# Examples:
#   i=6  -> 2^6  = 64 elements   -> 512 Bytes
#   i=20 -> 2^20 = 1M elements   -> 8 MB
#   i=24 -> 2^24 = 16M elements  -> 128 MB
SIZES = [2**i for i in range(6, 25)] 
NUM_POINTS_PER_SIZE = 100
DATA_RATE_S = 0.05  # seconds


class MinimalPublisher(Node):
    def __init__(self):
        super().__init__("minimal_publisher")
        self.publisher_ = self.create_publisher(UInt64MultiArray, "topic", 10)
        timer_period = DATA_RATE_S
        self.timer = self.create_timer(timer_period, self.timer_callback)
        self.i = 0
        self.j = 0

    def timer_callback(self):
        random_data = np.array(
            np.random.randint(255, size=SIZES[self.i], dtype=np.uint64)
        )
        random_data[0] = time.perf_counter_ns()

        msg = UInt64MultiArray(data=random_data.tolist())
        self.publisher_.publish(msg)
        # self.get_logger().info('Publishing: "%s"' % msg.data)
        if self.j == NUM_POINTS_PER_SIZE:
            self.i += 1
            self.j = 0
            if self.i >= len(SIZES):
                self.get_logger().info("Benchmarking data collection complete!")
                self.timer.cancel()
                self.destroy_node()
        else:
            self.j += 1


def main(args=None):
    rclpy.init(args=args)

    minimal_publisher = MinimalPublisher()

    try:
        rclpy.spin(minimal_publisher)
    except KeyboardInterrupt:
        pass
    finally:
        minimal_publisher.destroy_node()
        rclpy.shutdown()


if __name__ == "__main__":
    main()

File: experiments/benchmarks/ros/benchmark_cpp/src/publisher.cpp
#include <chrono>
#include <memory>
#include <random>
#include <string>

#include "rclcpp/rclcpp.hpp"
#include "rclcpp_components/register_node_macro.hpp"
#include "std_msgs/msg/u_int64_multi_array.hpp"

/**
 *  Publisher node for ROS 2 C++ benchmarking.
 */

// NOTE: kSizes defines the number of uint64 elements in the payload.
// Since each uint64 is 8 bytes, the actual payload size in bytes is: Element Count * 8.
// Matches SIZES in benchmark_python/publisher.py: [2^i for i in range(6, 25)].
//   i=6  -> 2^6  = 64 elements   -> 512 Bytes
//   i=20 -> 2^20 = 1M elements   -> 8 MB
//   i=24 -> 2^24 = 16M elements  -> 128 MB
std::vector<size_t> kSizes = []() {
    std::vector<size_t> sizes;
    sizes.reserve(25 - 6);
    for (size_t i = 6; i < 25; ++i) {
        sizes.push_back(static_cast<size_t>(1) << i);
    }
    return sizes;
}();
static constexpr size_t kNumPointsPerSize {100};
static constexpr std::chrono::milliseconds kDataRate {50};


class MinimalPublisher : public rclcpp::Node
{
public:
  explicit MinimalPublisher(const rclcpp::NodeOptions & options)
  : Node("minimal_publisher", options), i_(0), j_(0), rng_(std::random_device{}()), dist_(0, 254)
  {
    publisher_ = this->create_publisher<std_msgs::msg::UInt64MultiArray>("topic", 10);
    auto timer_callback =
      [this]() -> void {
        auto message = std::make_unique<std_msgs::msg::UInt64MultiArray>();
        message->data.resize(kSizes.at(i_));
        for (auto& elem : message->data)
        {
            elem = dist_(rng_);
        }
        message->data[0] = std::chrono::duration_cast<std::chrono::nanoseconds>(
            std::chrono::steady_clock::now().time_since_epoch()
        ).count();
        // RCLCPP_INFO(this->get_logger(), "Publishing message of size %lu", kSizes.at(i_));
        this->publisher_->publish(std::move(message));

        if (j_ == kNumPointsPerSize)
        {
            ++i_;
            j_ = 0;
            if (i_ >= kSizes.size())
            {
                RCLCPP_INFO(this->get_logger(), "Benchmarking data collection complete!");
                timer_->cancel();
            }
        } else {
            ++j_;
        }
      };
    timer_ = this->create_wall_timer(kDataRate, timer_callback);
  }

private:
  rclcpp::TimerBase::SharedPtr timer_;
  rclcpp::Publisher<std_msgs::msg::UInt64MultiArray>::SharedPtr publisher_;
  size_t i_;
  size_t j_;
  std::mt19937_64 rng_;
  std::uniform_int_distribution<uint64_t> dist_;
};

RCLCPP_COMPONENTS_REGISTER_NODE(MinimalPublisher)

#ifndef BENCHMARK_CPP_AS_COMPONENT
int main(int argc, char * argv[])
{
  rclcpp::init(argc, argv);
  rclcpp::spin(std::make_shared<MinimalPublisher>(rclcpp::NodeOptions()));
  rclcpp::shutdown();
  return 0;
}
#endif
